diff --git a/org/eclipse/osgi/internal/loader/classpath/ClasspathManager.java b/org/eclipse/osgi/internal/loader/classpath/ClasspathManager.java
index f7eedad982ffb..e4fc503bdefe5 100644
--- a/org/eclipse/osgi/internal/loader/classpath/ClasspathManager.java
+++ b/org/eclipse/osgi/internal/loader/classpath/ClasspathManager.java
@@ -20,8 +20,12 @@ import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Enumeration;
+import java.util.HashMap;
 import java.util.List;
 import java.util.ListIterator;
+import java.util.Map;
+import java.util.Properties;
+import java.util.function.BiFunction;
 import org.eclipse.osgi.container.Module;
 import org.eclipse.osgi.container.ModuleCapability;
 import org.eclipse.osgi.container.ModuleContainerAdaptor.ContainerEvent;
@@ -535,6 +539,15 @@ public class ClasspathManager {
 		String filename = name.replace('.', '/').concat(".class"); //$NON-NLS-1$
 
 		BundleEntry entry = classpathEntry.findEntry(filename);
+		if ((entry == null) && (_reverseTextReplacerBiFunction != null)) {
+			String newFilename = _reverseTextReplacerBiFunction.apply(
+				generation.getBundleFile() + "#", filename);
+
+			if (!newFilename.equals(filename)) {
+				entry = classpathEntry.findEntry(newFilename);
+			}
+		}
+
 		if (entry == null)
 			return null;
 
@@ -833,5 +846,34 @@ public class ClasspathManager {
 	public ModuleClassLoader getClassLoader() {
 		return classloader;
 	}
+
+	private static final BiFunction<String, String, String>
+		_reverseTextReplacerBiFunction;
+
+	static {
+		ClassLoader classLoader = ClassLoader.getSystemClassLoader();
+
+		Object instance = null;
+
+		try {
+			Class<?> clazz = classLoader.loadClass(
+				"com.liferay.portal.tools.jakarta.ee.transformer.function." +
+					"ReverseTextReplacerBiFunction");
+
+			instance = clazz.newInstance();
+		}
+		catch (ReflectiveOperationException reflectiveOperationException) {
+			if (!(reflectiveOperationException instanceof
+					ClassNotFoundException)) {
+
+				throw new ExceptionInInitializerError(
+					reflectiveOperationException);
+			}
+		}
+
+		_reverseTextReplacerBiFunction =
+			(BiFunction<String, String, String>)instance;
+	}
+
 }
 /* @generated */
diff --git a/org/eclipse/osgi/storage/BundleInfo.java b/org/eclipse/osgi/storage/BundleInfo.java
index e4ebfd822561a..cc75e7ca967bd 100644
--- a/org/eclipse/osgi/storage/BundleInfo.java
+++ b/org/eclipse/osgi/storage/BundleInfo.java
@@ -25,9 +25,11 @@ import java.util.Enumeration;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
+import java.util.Properties;
 import java.util.ResourceBundle;
 import java.util.Set;
 import java.util.concurrent.TimeUnit;
+import java.util.function.BiFunction;
 import org.eclipse.osgi.container.Module;
 import org.eclipse.osgi.container.ModuleContainerAdaptor.ModuleEvent;
 import org.eclipse.osgi.container.ModuleRevision;
@@ -148,7 +150,13 @@ public final class BundleInfo {
 									}
 								}
 							}
-							rawHeaders = Collections.unmodifiableMap(merged);
+
+							if (_bundleHeaderReplacerBiFunction == null) {
+								rawHeaders = Collections.unmodifiableMap(merged);
+							}
+							else {
+								rawHeaders = Collections.unmodifiableMap((Map<String, String>)(Map)_bundleHeaderReplacerBiFunction.apply(bundleFile + "#", (Map<Object, Object>)(Map)merged));
+							}
 						} catch (Exception e) {
 							if (e instanceof RuntimeException) {
 								throw (RuntimeException) e;
@@ -639,4 +647,35 @@ public final class BundleInfo {
 		}
 	}
 
+	private static final BiFunction
+		<String, Map<Object, Object>, Map<Object, Object>>
+			_bundleHeaderReplacerBiFunction;
+
+	static {
+		ClassLoader classLoader = ClassLoader.getSystemClassLoader();
+
+		Object instance = null;
+
+		try {
+			Class<?> clazz = classLoader.loadClass(
+				"com.liferay.portal.tools.jakarta.ee.transformer.function." +
+					"BundleHeaderReplacerBiFunction");
+
+			instance = clazz.newInstance();
+		}
+		catch (ReflectiveOperationException reflectiveOperationException) {
+			if (!(reflectiveOperationException instanceof
+					ClassNotFoundException)) {
+
+				throw new ExceptionInInitializerError(
+					reflectiveOperationException);
+			}
+		}
+
+		_bundleHeaderReplacerBiFunction =
+			(BiFunction<String, Map<Object, Object>, Map<Object, Object>>)
+				instance;
+	}
+
 }
+/* @generated */
\ No newline at end of file
